server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/WriteLoadConstraintDecider.java

```diff
@@ -117,9 +117,50 @@ public Decision canRemain(IndexMetadata indexMetadata, ShardRouting shardRouting
             return Decision.single(Decision.Type.YES, NAME, "canRemain() is not enabled");
         }
 
-        // TODO: implement
+        // Check whether the shard being relocated has any write load estimate. If it does not, then this decider has no opinion.
+        var allShardWriteLoads = allocation.clusterInfo().getShardWriteLoads();
+        var shardWriteLoad = allShardWriteLoads.get(shardRouting.shardId());
+        if (shardWriteLoad == null || shardWriteLoad == 0) {
+            return Decision.single(Decision.Type.YES, NAME, "Shard has no estimated write load. Decider takes no action.");
+        }
+
+        var allNodeUsageStats = allocation.clusterInfo().getNodeUsageStatsForThreadPools();
+        var nodeUsageStatsForThreadPools = allNodeUsageStats.get(node.nodeId());
+        if (nodeUsageStatsForThreadPools == null) {
+            // No node-level thread pool usage stats were reported for this node. Let's assume this is OK and that the simulator will handle
+            // setting a node-level write load for this node after this shard is assigned.
+            return Decision.single(Decision.Type.YES, NAME, "The node has no write load estimate. Decider takes no action.");
+        }
+
+        assert nodeUsageStatsForThreadPools.threadPoolUsageStatsMap().isEmpty() == false;
+        assert nodeUsageStatsForThreadPools.threadPoolUsageStatsMap().get(ThreadPool.Names.WRITE) != null;
+        var nodeWriteThreadPoolStats = nodeUsageStatsForThreadPools.threadPoolUsageStatsMap().get(ThreadPool.Names.WRITE);
+        var nodeWriteThreadPoolQueueLatencyThreshold = writeLoadConstraintSettings.getQueueLatencyThreshold();
+        if (nodeWriteThreadPoolStats.maxThreadPoolQueueLatencyMillis() >= nodeWriteThreadPoolQueueLatencyThreshold.millis()) {
+            String explain = Strings.format(
+                "Node [%s] has a queue latency of [%d] millis that exceeds the queue latency threshold of [%s]. This node is hot-spotting. "
+                    + "Current thread pool utilization [%f]. Moving shard(s) away.",
+                node.nodeId(),
+                nodeWriteThreadPoolStats.maxThreadPoolQueueLatencyMillis(),
+                nodeWriteThreadPoolQueueLatencyThreshold.toHumanReadableString(2),
+                nodeWriteThreadPoolStats.averageThreadPoolUtilization()
+            );
+            logger.debug(explain);
+            return Decision.single(Decision.Type.NOT_PREFERRED, NAME, explain);
+        }
 
-        return Decision.single(Decision.Type.YES, NAME, "canRemain() is not yet implemented");
+        String explanation = Strings.format(
+            "Node [%s]'s queue latency of [%d] does not exceed the threshold of [%s]",
+            node.nodeId(),
+            nodeWriteThreadPoolStats.maxThreadPoolQueueLatencyMillis(),
+            nodeWriteThreadPoolQueueLatencyThreshold.toHumanReadableString(2)
+        );
+
+        if (logger.isTraceEnabled()) {
+            logger.trace(explanation);
+        }
```

Comment:
Nit: I don't think we need to wrap this in the conditional, log4j will do that immediately inside `logger.trace(...)`. Explicit checks like this are only useful to avoid expensive pre-calculation or fetching that may be required for the log message. Here we're just passing in a string that already exists (i.e. zero cost).